src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,266 @@ def use_quotas(
             indicates how much quota should actually be consumed.
         """
         pass
+
+
+class RedisCardinalityLimiter(CardinalityLimiter):
+    """
+    The Redis cardinality limiter stores a key per unit hash, and adds the unit
+    hash to a set key prefixed with `RequestedQuota.prefix`. The memory usage
+    grows with the dimensionality of `prefix` and the `unit_hashes` that fit
+    into the limit.
+
+    Many design decisions for this cardinality limiter were copied from the
+    sliding_windows rate limiter. You will find the next couple of paragraphs
+    in its documentation as well.
+
+    Why is checking quotas and using quotas two separate implementations?
+    Isn't that a time-of-check-time-of-use bug, and allows me to over-spend
+    quota when requests are happening concurrently?
+
+    1) It's desirable to first check quotas, then do a potentially fallible
+       operation, then consume quotas. This rate limiter is primarily going to
+       be used inside of the metrics string indexer to rate-limit database
+       writes. What we want to do there is: read DB, check rate limits,
+       write to DB, use rate limits.
+
+       If we did it any other way (the obvious one being to read DB,
+       check-and-use rate limits, write DB), crashes while writing to the
+       database can over-consume quotas. This is not a big problem if those
+       crashes are flukes, and especially not a problem if the crashes are
+       a result of an overloaded DB.
+
+       It is however a problem in case the consumer is crash-looping, or
+       crashing (quickly) for 100% of requests (e.g. due to schema
+       mismatches between code and DB that somehow don't surface during the
+       DB read). In that case the quotas would be consumed immediately and
+       incident recovery would require us to reset all quotas manually (or
+       disable rate limiting via some killswitch)
+
+    3) The redis backend (really the only backend we care about) already
+       has some consistency problems.
+
+       a) Redis only provides strong consistency and ability to
+          check-and-increment counters when all involved keys hit the same
+          Redis node. That means that a quota with prefix="org_id:123" can
+          only run on a single redis node. It also means that a global
```

Comment:
We do not have counters here. The problem is not present.
I think this came from the siling window rate limiter. This docstring should be updated